core/src/test/java/org/apache/iceberg/catalog/CatalogTests.java

```diff
@@ -3167,6 +3172,412 @@ public void testRegisterExistingTable() {
+    SnapshotRef tagRef1 = refs.get(tagName1);
+    assertThat(tagRef1.snapshotId()).isEqualTo(secondSnapshot.snapshotId());
+    assertThat(tagRef1.snapshotId()).isNotEqualTo(firstSnapshot.snapshotId());
+
+    // Add more data to main branch
+    table.newFastAppend().appendFile(FILE_C).commit();
+    Snapshot thirdSnapshot = table.currentSnapshot();
+
+    // Verify tag still points to original snapshot (tags don't move with main branch)
+    refs = table.refs();
+    SnapshotRef mainRef = refs.get(SnapshotRef.MAIN_BRANCH);
+    SnapshotRef tagRef1Updated = refs.get(tagName1);
+
+    assertThat(mainRef.snapshotId()).isEqualTo(thirdSnapshot.snapshotId());
+    assertThat(tagRef1Updated.snapshotId()).isEqualTo(secondSnapshot.snapshotId());
+    assertThat(tagRef1Updated.snapshotId()).isNotEqualTo(mainRef.snapshotId());
+
+    table.manageSnapshots().removeTag(tagName2).commit();
+
+    // Verify tag is removed
+    refs = table.refs();
+    assertThat(refs).doesNotContainKey(tagName2);
+  }
+
+  @Test
+  public void testRetentionPolicyManagement() {
+    C catalog = catalog();
+
+    if (requiresNamespaceCreate()) {
+      catalog.createNamespace(NS);
+    }
+
+    Table table =
+        catalog.buildTable(TABLE, SCHEMA).withProperty(TableProperties.GC_ENABLED, "true").create();
+
+    table.newFastAppend().appendFile(FILE_A).commit();
+    Snapshot firstSnapshot = table.currentSnapshot();
+
+    String branchName = "retention-branch";
+    int minSnapshotsToKeep = 2;
+    long maxSnapshotAgeMs = 1000L;
+
+    table
+        .manageSnapshots()
+        .createBranch(branchName, firstSnapshot.snapshotId())
+        .setMinSnapshotsToKeep(branchName, minSnapshotsToKeep)
+        .setMaxSnapshotAgeMs(branchName, maxSnapshotAgeMs)
+        .commit();
+
+    table.newFastAppend().toBranch(branchName).appendFile(FILE_B).commit();
+    table.newFastAppend().toBranch(branchName).appendFile(FILE_C).commit();
+
+    // Wait for snapshots to age
+    Uninterruptibles.sleepUninterruptibly(1100, TimeUnit.MILLISECONDS);
+
+    // Expire snapshots
+    table.expireSnapshots().cleanExpiredMetadata(true).commit();
+
+    // Verify retention policy was enforced
+    Map<String, SnapshotRef> refs = table.refs();
+    SnapshotRef branchRef = refs.get(branchName);
+    assertThat(branchRef).isNotNull();
+
+    // The branch should still exist and point to a recent snapshot
+    Snapshot currentBranchSnapshot = table.snapshot(branchRef.snapshotId());
+    assertThat(currentBranchSnapshot).isNotNull();
+
+    String tagName = "short-lived-tag";
+    long maxRefAgeMs = 1000L; // 1 second
+
+    table
+        .manageSnapshots()
+        .createTag(tagName, firstSnapshot.snapshotId())
+        .setMaxRefAgeMs(tagName, maxRefAgeMs)
+        .commit();
+
+    // Verify tag exists
+    refs = table.refs();
+    assertThat(refs).containsKey(tagName);
+
+    // Wait for tag to age
+    Uninterruptibles.sleepUninterruptibly(1100, TimeUnit.MILLISECONDS);
```

Comment:
We have a helper, TestHelpers.waitUntilAfter, I'd prefer to use that for code consistency.